### PR TITLE
CONTRIB: Automate AUTHORS file update

### PR DESCRIPTION
**Pull request by tvegas1**
## What?
Automate AUTHORS updates according to documentation.

## Why?
Seems better to try to have common approach.

## How?
```shell
$ ./contrib/authors_update.sh upstream/v1.17.x..upstream/v1.18.x
..
   Artemy Kovalyov <artemyko@nvidia.com>
++ Arun Chandran <Arun.Chandran@amd.com>
   Edgar Gabriel <Edgar.Gabriel@amd.com>
   Evgeny Leksikov <evgenylek@mellanox.com>
..
```

### CODE REVIEW

**Review by yosefe — COMMENTED** [review]
Can we add this script to CI, so that if new author is adding a commit and it is not in AUTHORS file, the CI fails?

File: contrib/authors_update.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        echo "$line" >>$tmp
+        echo "++ $line"
+    else
+        echo "   $line"
+    fi
+done
+
+sort -o $tmp{,}
+grep -v @ AUTHORS >>$tmp
+mv $tmp AUTHORS
```
Comment (michal-shalev) [inline contrib/authors_update.sh:23]: add empty line at the end?

File: contrib/authors_update.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set -eu
+
+range="${1?Provide commit range like orig/v1.17..orig/v1.18}"
+
+tmp=AUTHORS.email
```
Comment (michal-shalev) [inline contrib/authors_update.sh:7]: maybe use mktemp?

File: contrib/authors_update.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set -eu
+
+range="${1?Provide commit range like orig/v1.17..orig/v1.18}"
+
```
Comment (michal-shalev) [inline contrib/authors_update.sh:6]: add a check that AUTHORS file exists?

File: contrib/authors_update.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+tmp=AUTHORS.email
+grep @ AUTHORS >$tmp
+
+git log --no-merges --pretty=format:"%an%x09%ae" "$range" | sort | uniq | \
```
Comment (michal-shalev) [inline contrib/authors_update.sh:10]: maybe check if `git log` produces any output? add a log?

File: contrib/authors_update.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set -eu
+
+range="${1?Provide commit range like orig/v1.17..orig/v1.18}"
```
Comment (michal-shalev) [inline contrib/authors_update.sh:5]: add a check that the range is valid? (not empty)

File: contrib/authors_update.sh (hunk 2 above)
Comment (tvegas1) [reply contrib/authors_update.sh]: fixed

File: contrib/authors_update.sh (hunk 1 above)
Comment (tvegas1) [reply contrib/authors_update.sh:28]: seems it shows in red with git diff

File: contrib/authors_update.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set -eu
```
[the remark this answers is not in the document]
Comment (tvegas1) [reply contrib/authors_update.sh:7]: fixed both

File: contrib/authors_update.sh (hunk 4 above)
Comment (tvegas1) [reply contrib/authors_update.sh:15]: fixed

File: contrib/authors_update.sh (hunk 5 above)
Comment (tvegas1) [reply contrib/authors_update.sh:10]: fixed with git log

**Comment by tvegas1** [comment]
adding to CI imposes PR creators to have their commits under exact email, or name, could this become a problem at some point?

File: contrib/authors_update.sh [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# See file LICENSE for terms.
+#
+
+set -eu
+set -o pipefail
+
+range="${1?Provide commit range like orig/v1.17..orig/v1.18}"
+
+if [ ! -f AUTHORS ]
```
Comment (michal-shalev) [inline contrib/authors_update.sh:12]: maybe also check `-w AUTHORS` (writeable?)

File: contrib/authors_update.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+then
+    echo "AUTHORS file not accessible"
+    exit 1
+fi
+
+# Failure triggered if range is not valid
+git --no-pager log --oneline --pretty=format:"%h %an <%ae> %s" "$range"
```
Comment (michal-shalev) [inline contrib/authors_update.sh:19]: maybe add a log if the range is not valid?

File: contrib/authors_update.sh (hunk 4 above)
Comment (michal-shalev) [reply contrib/authors_update.sh:27]: maybe add a log if it doesn't produces any output?

File: buildlib/pr/codestyle.yml [hunk 9]
```diff
@@ -63,6 +63,33 @@ jobs:
...
+    steps:
+      - checkout: self
+        clean: true
+        fetchDepth: 100
+        retryCountOnTaskFailure: 5
+
+      - bash: |
+          set -eE
+          set -x
+          source ./buildlib/az-helpers.sh
```
Comment (michal-shalev) [inline buildlib/pr/codestyle.yml:82]: why here and not inside the script? also, which function of `az-helpers.sh` are you using?

File: buildlib/pr/codestyle.yml [hunk 10]
```diff
@@ -63,6 +63,33 @@ jobs:
           fi
         condition: eq(variables['Build.Reason'], 'PullRequest')
 
+  - job: author
+    displayName: AUTHORS file check
```
Comment (michal-shalev) [inline buildlib/pr/codestyle.yml:67]: maybe include the word `update` in the name?

File: buildlib/pr/codestyle.yml [hunk 11]
```diff
@@ -63,6 +63,33 @@ jobs:
...
+    container: fedora
+    steps:
+      - checkout: self
+        clean: true
+        fetchDepth: 100
+        retryCountOnTaskFailure: 5
+
+      - bash: |
+          set -eE
+          set -x
```
Comment (michal-shalev) [inline buildlib/pr/codestyle.yml:81]: you can merge it to `set -eEx`

File: contrib/authors_update.sh [hunk 12]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# See file LICENSE for terms.
+#
+
+set -eu
+set -o pipefail
```
Comment (michal-shalev) [inline contrib/authors_update.sh:8]: can marge into `set -eEx -o pipefail`

File: contrib/authors_update.sh (hunk 8 above)
Comment (tvegas1) [reply contrib/authors_update.sh:19]: I let git output provide detailed:
```
/contrib/authors_update.sh bad..range
fatal: ambiguous argument 'bad..range': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

File: contrib/authors_update.sh (hunk 12 above)
Comment (tvegas1) [reply contrib/authors_update.sh:8]: fixed, but i'd prefer keep -x for CI only.

File: buildlib/pr/codestyle.yml (hunk 9 above)
Comment (tvegas1) [reply buildlib/pr/codestyle.yml:82]: removed

File: buildlib/pr/codestyle.yml (hunk 11 above)
Comment (tvegas1) [reply buildlib/pr/codestyle.yml:81]: fixed

File: buildlib/pr/codestyle.yml (hunk 10 above)
Comment (tvegas1) [reply buildlib/pr/codestyle.yml]: fixed

File: contrib/authors_update.sh (hunk 4 above)
Comment (tvegas1) [reply contrib/authors_update.sh:31]: fixed, but I only see something like "HEAD..HEAD" to do that..

File: contrib/authors_update.sh (hunk 12 above)
Comment (michal-shalev) [reply contrib/authors_update.sh]: IMO should add `E`

Comment (tvegas1) [reply contrib/authors_update.sh]: fixed

**Review by michal-shalev — APPROVED** [review]
Nice

File: contrib/authors_update.sh [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+tmp=$(mktemp --tmpdir=./)
+grep @ AUTHORS >"$tmp"
+
+echo
+echo "Names:"
+
+git log --no-merges --pretty=format:"%an%x09%ae" "$range" | sort | uniq | \
+while IFS=$'\t' read -r name email; do
+    line="$name <$email>"
+    if ! grep -iqw "$email" "$tmp" && ! grep -iq "$name" "$tmp"; then
```
[the remark this answers is not in the document]
Comment (tvegas1) [reply contrib/authors_update.sh:35]: added

File: contrib/authors_update.sh [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+if [ "$(git log --no-merges --oneline "$range" | wc -l)" -eq 0 ]
+then
+    echo "Error: empty range \"$range\""
+    exit 1
+fi
+
+# Failure triggered if range is not valid
+git --no-pager log --no-merges --oneline --pretty=format:"%h %an <%ae> %s" \
+    "$range"
```
[the remark this answers is not in the document]
Comment (tvegas1) [reply contrib/authors_update.sh:25]: ok make sense, retested empty/bad range and patched

File: contrib/authors_update.sh [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+# See file LICENSE for terms.
+#
+
+set -eEu -o pipefail
+
+range="${1?Provide commit range like orig/v1.17..orig/v1.18}"
+
+if [ ! -w AUTHORS ]
+then
+    echo "AUTHORS file not accessible"
```
Comment (gleon99) [inline contrib/authors_update.sh:13]: `is not writable`

File: contrib/authors_update.sh [hunk 16]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# See file LICENSE for terms.
+#
+
+set -eEu -o pipefail
+
+range="${1?Provide commit range like orig/v1.17..orig/v1.18}"
```
Comment (gleon99) [inline contrib/authors_update.sh:9]: `like` -> `, for example:`

File: contrib/authors_update.sh [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+then
+    echo "AUTHORS file not accessible"
+    exit 1
+fi
+
+# Failure message triggered if range is not valid
+lines=$(git log --no-merges --pretty=format:"%an%x09%ae" "$range" | sort | uniq)
+if [ -z "$lines" ]
+then
+    echo "Error: empty range \"$range\""
```
Comment (gleon99) [inline contrib/authors_update.sh:21]: `provided range is empty`

File: contrib/authors_update.sh (hunk 15 above)
Comment (tvegas1) [reply contrib/authors_update.sh:13]: fixed, but -w also covers the case where file does not exist

File: contrib/authors_update.sh (hunk 16 above)
Comment (tvegas1) [reply contrib/authors_update.sh:9]: fixed

File: contrib/authors_update.sh (hunk 17 above)
Comment (tvegas1) [reply contrib/authors_update.sh:21]: fixed

File: contrib/authors_update.sh [hunk 18]
```diff
@@ -6,32 +6,32 @@
...
 if [ ! -w AUTHORS ]
 then
-    echo "AUTHORS file not accessible"
+    echo "AUTHORS file is not writable"
     exit 1
 fi
 
-if [ "$(git log --no-merges --oneline "$range" | wc -l)" -eq 0 ]
+# Failure message triggered if range is not valid
+lines=$(git log --no-merges --pretty=format:"%an%x09%ae" "$range" | sort | uniq)
```
Comment (yosefe) [inline contrib/authors_update.sh:18]: minor: add "-u" flag to sort instead of pipe to uniq

Comment (tvegas1) [reply contrib/authors_update.sh]: fixed

**Review by yosefe — APPROVED** [review]

**Comment by michal-shalev** [comment]
image
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=92367&view=logs&jobId=d572d9ce-6832-5171-db6a-7a9074180360&j=d572d9ce-6832-5171-db6a-7a9074180360&t=f889154a-ca50-5f7a-dec3-4e4d8a99ad41
@tvegas1 @yosefe 
Can you please check?

**Pull request merged**